File: mobile/android/MeowCrypto.kt
/**
 * MeowCrypto.kt - Android Stub Implementation
 * 🐱 Meow Decoder Android Cryptographic Wrapper
 *
 * STATUS: STUB - Not yet connected to Rust crypto_core
 *
 * This file defines the API surface for the Android crypto wrapper.
 * Actual implementation will use JNI bindings to the Rust crypto_core.
 */

package io.github.meowdecoder.crypto

import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec

/**
 * Errors that can occur during cryptographic operations
 */
sealed class MeowCryptoError : Exception() {
    object InvalidKeyLength : MeowCryptoError()
    object InvalidNonceLength : MeowCryptoError()
    object InvalidSaltLength : MeowCryptoError()
    object EncryptionFailed : MeowCryptoError()
    object DecryptionFailed : MeowCryptoError()
    object AuthenticationFailed : MeowCryptoError()
    object KeyDerivationFailed : MeowCryptoError()
    object NotImplemented : MeowCryptoError()
}

/**
 * Result type for crypto operations
 */
sealed class MeowCryptoResult<out T> {
    data class Success<T>(val value: T) : MeowCryptoResult<T>()
    data class Error(val error: MeowCryptoError) : MeowCryptoResult<Nothing>()

    inline fun <R> fold(
        onSuccess: (T) -> R,
        onError: (MeowCryptoError) -> R
    ): R = when (this) {
        is Success -> onSuccess(value)
        is Error -> onError(error)
    }
}

/**
 * Main cryptographic interface for Meow Decoder
 *
 * This is a STUB implementation using Android's javax.crypto.
 * Production version will use the Rust crypto_core via JNI.
 */
object MeowCrypto {

    // MARK: - Constants

    /** Expected key length in bytes (256 bits) */
    const val KEY_LENGTH = 32

    /** Expected nonce length in bytes (96 bits) */
    const val NONCE_LENGTH = 12

    /** Expected salt length in bytes (128 bits) */
    const val SALT_LENGTH = 16

    /** GCM authentication tag length in bits */
    const val TAG_LENGTH_BITS = 128

    /** GCM authentication tag length in bytes */
    const val TAG_LENGTH = 16

    /** AES/GCM transformation string */
    private const val AES_GCM_TRANSFORMATION = "AES/GCM/NoPadding"

    /** Secure random instance */
    private val secureRandom = SecureRandom()

    // MARK: - Key Derivation

    /**
     * Derive encryption key from password using PBKDF2
     *
     * NOTE: STUB - Uses PBKDF2. Production will use Argon2id via JNI.
     *
     * @param password User password
     * @param salt Random salt (16 bytes)
     * @return Derived key (32 bytes) wrapped in Result
     */
    fun deriveKey(password: String, salt: ByteArray): MeowCryptoResult<ByteArray> {
        if (salt.size != SALT_LENGTH) {
            return MeowCryptoResult.Error(MeowCryptoError.InvalidSaltLength)
        }

        return try {
            // STUB: Using PBKDF2 as placeholder
            // Production: Call into Rust argon2_derive() via JNI
            val spec = PBEKeySpec(password.toCharArray(), salt, 100_000, KEY_LENGTH * 8)
            val factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
            val key = factory.generateSecret(spec).encoded
            spec.clearPassword()
            MeowCryptoResult.Success(key)
        } catch (e: Exception) {
            MeowCryptoResult.Error(MeowCryptoError.KeyDerivationFailed)
        }
    }

    // MARK: - Encryption

    /**
     * Encrypt data using AES-256-GCM
     *
     * @param plaintext Data to encrypt
     * @param key 32-byte encryption key
     * @param nonce 12-byte nonce (must be unique per encryption)
     * @param aad Optional additional authenticated data
     * @return Ciphertext with appended authentication tag
     */
    fun encrypt(
        plaintext: ByteArray,
        key: ByteArray,
        nonce: ByteArray,
        aad: ByteArray? = null
    ): MeowCryptoResult<ByteArray> {
        if (key.size != KEY_LENGTH) {
            return MeowCryptoResult.Error(MeowCryptoError.InvalidKeyLength)
        }
        if (nonce.size != NONCE_LENGTH) {
            return MeowCryptoResult.Error(MeowCryptoError.InvalidNonceLength)
        }

        return try {
            val cipher = Cipher.getInstance(AES_GCM_TRANSFORMATION)
            val keySpec = SecretKeySpec(key, "AES")
            val gcmSpec = GCMParameterSpec(TAG_LENGTH_BITS, nonce)

            cipher.init(Cipher.ENCRYPT_MODE, keySpec, gcmSpec)
            aad?.let { cipher.updateAAD(it) }

            val ciphertext = cipher.doFinal(plaintext)
            MeowCryptoResult.Success(ciphertext)
        } catch (e: Exception) {
            MeowCryptoResult.Error(MeowCryptoError.EncryptionFailed)
        }
    }

    /**
     * Decrypt data using AES-256-GCM
     *
     * @param ciphertext Encrypted data with tag appended
     * @param key 32-byte encryption key
     * @param nonce 12-byte nonce used during encryption
     * @param aad Optional additional authenticated data (must match encryption)
     * @return Decrypted plaintext
     */
    fun decrypt(
        ciphertext: ByteArray,
        key: ByteArray,
        nonce: ByteArray,
        aad: ByteArray? = null
    ): MeowCryptoResult<ByteArray> {
        if (key.size != KEY_LENGTH) {
            return MeowCryptoResult.Error(MeowCryptoError.InvalidKeyLength)
        }
        if (nonce.size != NONCE_LENGTH) {
            return MeowCryptoResult.Error(MeowCryptoError.InvalidNonceLength)
        }
        if (ciphertext.size < TAG_LENGTH) {
            return MeowCryptoResult.Error(MeowCryptoError.DecryptionFailed)
        }

        return try {
            val cipher = Cipher.getInstance(AES_GCM_TRANSFORMATION)
            val keySpec = SecretKeySpec(key, "AES")
            val gcmSpec = GCMParameterSpec(TAG_LENGTH_BITS, nonce)

            cipher.init(Cipher.DECRYPT_MODE, keySpec, gcmSpec)
            aad?.let { cipher.updateAAD(it) }

            val plaintext = cipher.doFinal(ciphertext)
            MeowCryptoResult.Success(plaintext)
        } catch (e: javax.crypto.AEADBadTagException) {
            MeowCryptoResult.Error(MeowCryptoError.AuthenticationFailed)
        } catch (e: Exception) {
            MeowCryptoResult.Error(MeowCryptoError.DecryptionFailed)
        }
    }

    // MARK: - Random Generation

    /**
     * Generate cryptographically secure random nonce
     *
     * @return 12-byte random nonce
     */
    fun generateNonce(): ByteArray {
        val bytes = ByteArray(NONCE_LENGTH)
        secureRandom.nextBytes(bytes)
        return bytes
    }

    /**
     * Generate cryptographically secure random salt
     *
     * @return 16-byte random salt
     */
    fun generateSalt(): ByteArray {
        val bytes = ByteArray(SALT_LENGTH)
        secureRandom.nextBytes(bytes)
        return bytes
    }

    /**
     * Generate cryptographically secure random key
     *
     * @return 32-byte random key
     */
    fun generateKey(): ByteArray {
        val bytes = ByteArray(KEY_LENGTH)
        secureRandom.nextBytes(bytes)
        return bytes
    }

    // MARK: - HMAC

    /**
     * Compute HMAC-SHA256
     *
     * @param key HMAC key
     * @param message Message to authenticate
     * @return 32-byte HMAC tag
     */
    fun hmacSha256(key: ByteArray, message: ByteArray): MeowCryptoResult<ByteArray> {
        return try {
            val mac = Mac.getInstance("HmacSHA256")
            val keySpec = SecretKeySpec(key, "HmacSHA256")
            mac.init(keySpec)
            val result = mac.doFinal(message)
            MeowCryptoResult.Success(result)
        } catch (e: Exception) {
            MeowCryptoResult.Error(MeowCryptoError.AuthenticationFailed)
        }
    }

    // MARK: - Constant Time Comparison

    /**
     * Compare two byte arrays in constant time
     *
     * Important: This prevents timing attacks
     *
     * @param a First value
     * @param b Second value
     * @return true if equal
     */
    fun constantTimeEqual(a: ByteArray, b: ByteArray): Boolean {
        if (a.size != b.size) return false

        var result: Int = 0
        for (i in a.indices) {
            result = result or (a[i].toInt() xor b[i].toInt())
        }

        return result == 0
    }

    // MARK: - Secure Memory

    /**
     * Securely zero memory
     *
     * Note: JVM makes true secure zeroing difficult.
     * Production should use Rust crypto_core which uses zeroize crate.
     *
     * @param data ByteArray to zero
     */
    fun secureZero(data: ByteArray) {
        java.util.Arrays.fill(data, 0.toByte())
    }
}

// MARK: - Extension Functions

/**
 * Convert ByteArray to hex string
 */
fun ByteArray.toHexString(): String = joinToString("") { "%02x".format(it) }

/**
 * Convert hex string to ByteArray
 */
fun String.hexToByteArray(): ByteArray {
    check(length % 2 == 0) { "Hex string must have even length" }
    return chunked(2)
        .map { it.toInt(16).toByte() }
        .toByteArray()
}

// MARK: - Keystore Integration Stub

/**
 * Android Keystore wrapper for secure key storage
 *
 * STATUS: STUB - Defines API for secure hardware-backed key storage
 *
 * Production implementation will:
 * - Use Android Keystore for key storage
 * - Use StrongBox if available (hardware security module)
 * - Require biometric authentication for key access
 */
object MeowKeystore {

    private const val KEYSTORE_PROVIDER = "AndroidKeyStore"
    private const val KEY_ALIAS_PREFIX = "meow_decoder_"

    /**
     * Check if StrongBox hardware security module is available
     *
     * @return true if StrongBox is available
     */
    fun isStrongBoxAvailable(): Boolean {
        // STUB: Would check Build.VERSION.SDK_INT >= Build.VERSION_CODES.P
        // and KeyInfo.isInsideSecureHardware()
        return false
    }

    /**
     * Generate and store a key in the Android Keystore
     *
     * @param alias Key alias
     * @param requireBiometric Require biometric authentication
     * @return true on success
     */
    fun generateKey(alias: String, requireBiometric: Boolean = false): Boolean {
        // STUB: Would use KeyGenerator with AndroidKeyStore provider
        // and KeyGenParameterSpec.Builder for key properties
        return false
    }

    /**
     * Retrieve a key from the Android Keystore
     *
     * @param alias Key alias
     * @return Key bytes or null if not found
     */
    fun getKey(alias: String): ByteArray? {
        // STUB: Would retrieve from KeyStore.getInstance(KEYSTORE_PROVIDER)
        return null
    }

    /**
     * Delete a key from the Android Keystore
     *
     * @param alias Key alias
     * @return true on success
     */
    fun deleteKey(alias: String): Boolean {
        // STUB: Would delete from KeyStore.getInstance(KEYSTORE_PROVIDER)
        return false
    }
}

// MARK: - Example Usage

/**
 * Example demonstrating MeowCrypto usage
 */
fun main() {
    println("🐱 MeowCrypto Android Stub Demo")
    println("=" .repeat(50))

    // Generate random values
    val salt = MeowCrypto.generateSalt()
    val nonce = MeowCrypto.generateNonce()
    println("Salt: ${salt.toHexString()}")
    println("Nonce: ${nonce.toHexString()}")

    // Derive key from password
    val password = "correct horse battery staple"
    val keyResult = MeowCrypto.deriveKey(password, salt)

    keyResult.fold(
        onSuccess = { key ->
            println("Derived key: ${key.toHexString()}")

            // Encrypt
            val plaintext = "Hello, Meow Decoder! 🐱".toByteArray(Charsets.UTF_8)
            val encryptResult = MeowCrypto.encrypt(plaintext, key, nonce)

            encryptResult.fold(
                onSuccess = { ciphertext ->
                    println("Ciphertext: ${ciphertext.toHexString()}")

                    // Decrypt
                    val decryptResult = MeowCrypto.decrypt(ciphertext, key, nonce)

                    decryptResult.fold(
                        onSuccess = { decrypted ->
                            println("Decrypted: ${String(decrypted, Charsets.UTF_8)}")
                            println("\n✅ Roundtrip successful!")
                        },
                        onError = { error ->
                            println("❌ Decryption failed: $error")
                        }
                    )
                },
                onError = { error ->
                    println("❌ Encryption failed: $error")
                }
            )

            // Secure cleanup
            MeowCrypto.secureZero(key)
        },
        onError = { error ->
            println("❌ Key derivation failed: $error")
        }
    )
}

File: mobile/ios/MeowCrypto.swift
// MeowCrypto.swift - iOS Stub Implementation
// 🐱 Meow Decoder iOS Cryptographic Wrapper
//
// STATUS: STUB - Not yet connected to Rust crypto_core
//
// This file defines the API surface for the iOS crypto wrapper.
// Actual implementation will use C-FFI bindings to the Rust crypto_core.

import Foundation
import CryptoKit

/// Errors that can occur during cryptographic operations
public enum MeowCryptoError: Error {
    case invalidKeyLength
    case invalidNonceLength
    case invalidSaltLength
    case encryptionFailed
    case decryptionFailed
    case authenticationFailed
    case keyDerivationFailed
    case notImplemented
}

/// Main cryptographic interface for Meow Decoder
///
/// This is a STUB implementation using Apple's CryptoKit.
/// Production version will use the Rust crypto_core via FFI.
public final class MeowCrypto {
    
    // MARK: - Constants
    
    /// Expected key length in bytes (256 bits)
    public static let keyLength = 32
    
    /// Expected nonce length in bytes (96 bits)
    public static let nonceLength = 12
    
    /// Expected salt length in bytes (128 bits)
    public static let saltLength = 16
    
    /// GCM authentication tag length in bytes
    public static let tagLength = 16
    
    // MARK: - Singleton
    
    /// Shared instance
    public static let shared = MeowCrypto()
    
    private init() {}
    
    // MARK: - Key Derivation
    
    /// Derive encryption key from password using Argon2id
    ///
    /// - Parameters:
    ///   - password: User password
    ///   - salt: Random salt (16 bytes)
    /// - Returns: Derived key (32 bytes)
    /// - Throws: MeowCryptoError if derivation fails
    ///
    /// - Note: STUB - Uses PBKDF2. Production will use Argon2id via FFI.
    public func deriveKey(password: String, salt: Data) throws -> Data {
        guard salt.count == Self.saltLength else {
            throw MeowCryptoError.invalidSaltLength
        }
        
        // STUB: Using PBKDF2 as placeholder
        // Production: Call into Rust argon2_derive()
        guard let passwordData = password.data(using: .utf8) else {
            throw MeowCryptoError.keyDerivationFailed
        }
        
        let key = try deriveKeyPBKDF2(
            password: passwordData,
            salt: salt,
            iterations: 100_000,
            keyLength: Self.keyLength
        )
        
        return key
    }
    
    /// PBKDF2 key derivation (STUB - placeholder for Argon2id)
    private func deriveKeyPBKDF2(
        password: Data,
        salt: Data,
        iterations: Int,
        keyLength: Int
    ) throws -> Data {
        var derivedKey = Data(count: keyLength)
        
        let status = derivedKey.withUnsafeMutableBytes { derivedKeyBytes in
            password.withUnsafeBytes { passwordBytes in
                salt.withUnsafeBytes { saltBytes in
                    CCKeyDerivationPBKDF(
                        CCPBKDFAlgorithm(kCCPBKDF2),
                        passwordBytes.baseAddress?.assumingMemoryBound(to: Int8.self),
                        password.count,
                        saltBytes.baseAddress?.assumingMemoryBound(to: UInt8.self),
                        salt.count,
                        CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA256),
                        UInt32(iterations),
                        derivedKeyBytes.baseAddress?.assumingMemoryBound(to: UInt8.self),
                        keyLength
                    )
                }
            }
        }
        
        guard status == kCCSuccess else {
            throw MeowCryptoError.keyDerivationFailed
        }
        
        return derivedKey
    }
    
    // MARK: - Encryption
    
    /// Encrypt data using AES-256-GCM
    ///
    /// - Parameters:
    ///   - plaintext: Data to encrypt
    ///   - key: 32-byte encryption key
    ///   - nonce: 12-byte nonce (must be unique per encryption)
    ///   - aad: Optional additional authenticated data
    /// - Returns: Ciphertext with appended authentication tag
    /// - Throws: MeowCryptoError if encryption fails
    public func encrypt(
        plaintext: Data,
        key: Data,
        nonce: Data,
        aad: Data? = nil
    ) throws -> Data {
        guard key.count == Self.keyLength else {
            throw MeowCryptoError.invalidKeyLength
        }
        guard nonce.count == Self.nonceLength else {
            throw MeowCryptoError.invalidNonceLength
        }
        
        // Use CryptoKit for encryption
        let symmetricKey = SymmetricKey(data: key)
        let nonceValue = try AES.GCM.Nonce(data: nonce)
        
        let sealedBox: AES.GCM.SealedBox
        if let aad = aad {
            sealedBox = try AES.GCM.seal(
                plaintext,
                using: symmetricKey,
                nonce: nonceValue,
                authenticating: aad
            )
        } else {
            sealedBox = try AES.GCM.seal(
                plaintext,
                using: symmetricKey,
                nonce: nonceValue
            )
        }
        
        // Return ciphertext + tag
        return sealedBox.ciphertext + sealedBox.tag
    }
    
    /// Decrypt data using AES-256-GCM
    ///
    /// - Parameters:
    ///   - ciphertext: Encrypted data with tag appended
    ///   - key: 32-byte encryption key
    ///   - nonce: 12-byte nonce used during encryption
    ///   - aad: Optional additional authenticated data (must match encryption)
    /// - Returns: Decrypted plaintext
    /// - Throws: MeowCryptoError if decryption or authentication fails
    public func decrypt(
        ciphertext: Data,
        key: Data,
        nonce: Data,
        aad: Data? = nil
    ) throws -> Data {
        guard key.count == Self.keyLength else {
            throw MeowCryptoError.invalidKeyLength
        }
        guard nonce.count == Self.nonceLength else {
            throw MeowCryptoError.invalidNonceLength
        }
        guard ciphertext.count >= Self.tagLength else {
            throw MeowCryptoError.decryptionFailed
        }
        
        let symmetricKey = SymmetricKey(data: key)
        let nonceValue = try AES.GCM.Nonce(data: nonce)
        
        // Split ciphertext and tag
        let ciphertextOnly = ciphertext.prefix(ciphertext.count - Self.tagLength)
        let tag = ciphertext.suffix(Self.tagLength)
        
        let sealedBox = try AES.GCM.SealedBox(
            nonce: nonceValue,
            ciphertext: ciphertextOnly,
            tag: tag
        )
        
        let plaintext: Data
        if let aad = aad {
            plaintext = try AES.GCM.open(sealedBox, using: symmetricKey, authenticating: aad)
        } else {
            plaintext = try AES.GCM.open(sealedBox, using: symmetricKey)
        }
        
        return plaintext
    }
    
    // MARK: - Random Generation
    
    /// Generate cryptographically secure random nonce
    ///
    /// - Returns: 12-byte random nonce
    public func generateNonce() -> Data {
        var bytes = [UInt8](repeating: 0, count: Self.nonceLength)
        _ = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
        return Data(bytes)
    }
    
    /// Generate cryptographically secure random salt
    ///
    /// - Returns: 16-byte random salt
    public func generateSalt() -> Data {
        var bytes = [UInt8](repeating: 0, count: Self.saltLength)
        _ = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
        return Data(bytes)
    }
    
    // MARK: - HMAC
    
    /// Compute HMAC-SHA256
    ///
    /// - Parameters:
    ///   - key: HMAC key
    ///   - message: Message to authenticate
    /// - Returns: 32-byte HMAC tag
    public func hmacSha256(key: Data, message: Data) -> Data {
        let symmetricKey = SymmetricKey(data: key)
        let mac = HMAC<SHA256>.authenticationCode(for: message, using: symmetricKey)
        return Data(mac)
    }
    
    // MARK: - Constant Time Comparison
    
    /// Compare two Data values in constant time
    ///
    /// - Parameters:
    ///   - a: First value
    ///   - b: Second value
    /// - Returns: true if equal
    ///
    /// - Important: This prevents timing attacks
    public func constantTimeEqual(_ a: Data, _ b: Data) -> Bool {
        guard a.count == b.count else { return false }
        
        var result: UInt8 = 0
        for i in 0..<a.count {
            result |= a[i] ^ b[i]
        }
        
        return result == 0
    }
    
    // MARK: - Secure Memory
    
    /// Securely zero memory (best-effort in Swift)
    ///
    /// - Parameter data: Data to zero
    ///
    /// - Note: Swift's memory model makes true secure zeroing difficult.
    /// Production should use Rust crypto_core which uses zeroize crate.
    public func secureZero(_ data: inout Data) {
        data.withUnsafeMutableBytes { bytes in
            memset_s(bytes.baseAddress, bytes.count, 0, bytes.count)
        }
    }
}

// MARK: - CommonCrypto Import for PBKDF2

import CommonCrypto

// MARK: - Extension for Hex Encoding

extension Data {
    /// Hex string representation
    public var hexString: String {
        map { String(format: "%02x", $0) }.joined()
    }
    
    /// Initialize from hex string
    public init?(hexString: String) {
        let len = hexString.count / 2
        var data = Data(capacity: len)
        
        var i = hexString.startIndex
        for _ in 0..<len {
            let j = hexString.index(i, offsetBy: 2)
            guard let byte = UInt8(hexString[i..<j], radix: 16) else {
                return nil
            }
            data.append(byte)
            i = j
        }
        
        self = data
    }
}
